parser_library/src/semantics/range_provider.cpp

```diff
@@ -169,3 +168,18 @@ range_provider::range_provider()
     : original_range()
     , state(adjusting_state::NONE)
 {}
+
+range text_range(
+    const lexing::logical_line::const_iterator& b, const lexing::logical_line::const_iterator& e, size_t lineno_offset)
+{
+    const auto [bx, by] = b.get_coordinates();
+    position b_pos(by + lineno_offset, bx);
+    if (b == e) // empty range
+        return range(std::move(b_pos));
+    else
+    {
+        const auto [ex, ey] = e.get_coordinates();
```

Comment:
The std::prev(e) should always be here, if your range happens to end at the end of segment, the end iterator points to the next (non-empty) line and the line number will be off.

At the same time my original suggestions did not cover the "empty logical line" case. So, I suggest the past-the-end iterator having coordinates [0,0] by definition i.e.
```
if (m_col == column_iterator()) return {0,0};
```
That way the function is always defined and "by accident" it handles the empty LL case correctly.